hazelcast/proxy/vector_collection.py

```diff
@@ -0,0 +1,282 @@
+import copy
+from typing import Any, Dict, List, Optional, Union, Tuple
+
+from hazelcast.future import Future, ImmediateFuture, combine_futures
+from hazelcast.protocol.codec import (
+    vector_collection_set_codec,
+    vector_collection_get_codec,
+    vector_collection_search_near_vector_codec,
+    vector_collection_delete_codec,
+    vector_collection_put_codec,
+    vector_collection_put_if_absent_codec,
+    vector_collection_remove_codec,
+    vector_collection_put_all_codec,
+)
+from hazelcast.proxy import Proxy
+from hazelcast.serialization.compact import SchemaNotReplicatedError
+from hazelcast.serialization.data import Data
+from hazelcast.util import check_not_none
+from hazelcast.vector import (
+    Document,
+    SearchResult,
+    Vector,
+    VectorPair,
+    VectorType,
+    VectorSearchOptions,
+)
+
+
+class VectorCollection(Proxy["BlockingVectorCollection"]):
+    def __init__(self, service_name, name, context):
+        super(VectorCollection, self).__init__(service_name, name, context)
+        self._reference_id_generator = context.lock_reference_id_generator
+
+    def blocking(self) -> "BlockingVectorCollection":
+        return BlockingVectorCollection(self)
+
+    def get(self, key: Any) -> Future[Optional[Document]]:
+        check_not_none(key, "key can't be None")
+        return self._get_internal(key)
+
+    def set(self, key: Any, document: Document) -> Future[None]:
+        check_not_none(key, "key can't be None")
+        check_not_none(document, "document can't be None")
+        check_not_none(document.value, "document value can't be None")
+        return self._set_internal(key, document)
+
+    def put(self, key: Any, document: Document) -> Future[Optional[Document]]:
+        check_not_none(key, "key can't be None")
+        check_not_none(document, "document can't be None")
+        check_not_none(document.value, "document value can't be None")
+        return self._put_internal(key, document)
+
+    def put_all(self, map: Dict[Any, Document]) -> Future[None]:
+        check_not_none(map, "map can't be None")
+        if not map:
+            return ImmediateFuture(None)
+        partition_service = self._context.partition_service
+        partition_map: Dict[str, List[Tuple[Data, Document]]] = {}
+
+        for key, doc in map.items():
+            check_not_none(key, "key can't be None")
+            check_not_none(doc, "value can't be None")
+            doc = copy.copy(doc)
+            doc.value = self._to_data(doc.value)
+            try:
+                entry = (self._to_data(key), doc)
+            except SchemaNotReplicatedError as e:
+                return self._send_schema_and_retry(e, self.put_all, map)
+
+            partition_id = partition_service.get_partition_id(entry[0])
+            try:
+                partition_map[partition_id].append(entry)
+            except KeyError:
```

Comment:
actually this key error is costly, we should use an explicit if statement here for key existence check

the reason is that if exception is triggers, try/except statements are slower than explicit if statements, see https://stackoverflow.com/a/2522013/21388077

The exception will be raised for sure and for each partition id and multiply that with the number of items in the map.items().

Reply:
Exceptions are not slow in Python.
We need to benchmark it before concluding the current way is slower.
But I changed that code to be more Pythonic at:
https://github.com/hazelcast/hazelcast-python-client/pull/674/commits/304e76ee18d663a8430d64e3de8b2206f09c6d89

Reply:
The stackoverflow answers may not apply universally.
We would have to benchmark different approaches, which I think is out-of-scope in this case.